examples/ctc.dx

```diff
@@ -0,0 +1,134 @@
+However, as mentioned in the original paper, we should only
+need the last step of the forward pass to compute the marginal
+probability, and Dex's autodiff should produce the backward
+pass automatically.  That makes this code much shorter than
+most implementations.
+
+
+def interleave (blank:v) (labels: m=>v) : (m & (Fin 2))=>v =
+  -- Turns "text" into "t e x t " by first pairing each letter with a blank,
+  -- then flattening the pairs back into a single-index table.
+  pairs = for i. [labels.i, blank]
+  for (i, j). pairs.i.j
+
+def prepend (first: v) (seq: m=>v) : (Unit|m)=>v =
+  -- Concatenates a single element to the beginning of a sequence.
+  for idx. case idx
+    Left () -> first
+    Right i -> seq.i
+
+def prepend_and_interleave (blank:v) (seq: m=>v) : 
+  ((Unit | (m & (Fin 2)))=>v) =
+  -- Turns "text" into " t e x t".
+  -- The output of this function has a slightly complicated output type, which
+  -- has size 1 + 2 * (size m).
+  interleaved = interleave blank seq
+  prepend blank interleaved
+
+def clipidx (n:Type) -> (i:Int) : n =
+  -- Returns element at 0 if less than zero.
+  -- Ideally we could have an alternative
+  -- to Fin that just clips the index at its bounds.
+  fromOrdinal n (select (i < 0) 0 i)
+
+def logaddexp (x:Real) (y:Real) : Real =
+  m = max x y
+  m + ( log ( (exp (x - m) + exp (y - m))))
+
+def ctc (time : Type) ?-> (vocab : Type) ?-> (position : Type) ?->
+        (dict: Eq vocab) ?=> (dict2: Eq position) ?=> (dict3: Eq time) ?=> (blank: vocab)
+        (logits: time=>vocab=>Real) (labels: position=>vocab) : Real =
+  -- Computes log p(labels | logits), marginalizing over possible alignments.
+  -- Todo: remove unnecessary implicit type annotations once
+  -- Dex starts putting implicit types in scope.
+
+  ilabels = prepend_and_interleave blank labels
+
+  normalized_logits = for t. logsoftmax logits.t
+
+  -- Initialization rules
+  logprob_start_with_blank = normalized_logits.(0@_).blank
+  logprob_of_first_label   = normalized_logits.(0@_).(labels.(0@_))
+  log_prob_seq_t0 = for pos.
+    select  (pos == (0@_)) logprob_start_with_blank (select (pos == (1@_)) logprob_of_first_label (log 0.000001))
+
+  same_as_last = \ilabels s.
+    o = ordinal s
+    select (o >= 2) (ilabels.s == ilabels.(clipidx _ (o - 2))) False
+
+  safe_idx = \prev s.
+    select (s >= 0) prev.(clipidx _ s) (log 0.0)
+
+  -- Todo: As suggested by Adam Paske, we could get rid of these
+  -- logaddexp calls with a newtype that overloads + and *
+  update = \t prev.
+    select (t == (0@_)) log_prob_seq_t0 for s.
+      cond = ilabels.s == blank || same_as_last ilabels s
+      labar = logaddexp prev.s (safe_idx prev ((ordinal s) - 1))
+      other = logaddexp labar  (safe_idx prev ((ordinal s) - 2))
+      ans = select cond labar other
+      ans + normalized_logits.t.(ilabels.s)
+
+  log_prob_seq_final = fold log_prob_seq_t0 update
+
+  -- Todo: nicer way to get last two elements of log_prob_seq_final.
```

Comment:
We should probably just add functions like `firstIdx : Ix n ?=> n` and `lastIdx : Ix n ?=> n`. Unfortunately, these are partial functions because the index sets might be empty.

Reply:
We might be able to address this issue with more typeclasses, e.g. ```firstIdx``` and ```lastIdx``` might be entries of a ```nonEmptyIx``` dictionary.

Reply:
Yeah, but it would be hard to prove that a type really should have `nonEmptyIx` statically. Certainly `Fin n`, for a run-time computed `n` doesn't have it, and neither do most range types.

Reply:
Maybe, but I'd love to go down this road a little.  It does have a Turing-tarpit flavor, but now that Dex has dependent types I wonder if most of these guarantees couldn't be passed around mostly automatically.  For instance, in this demo, the padded text already has a complicated type that does guarantee it's non-empty.  But pleasingly, I can still iterate over it without spelling this complicated type.  So maybe these extra proofs / guarantees could ride along with the type implicitly in most user code.

Reply:
We could certainly expand the dictionary-synthesis code to do some more elaborate proof search, but it will not be able to handle all possible use cases. E.g. `Fin n` for `n = 1 + abs x` is always non-empty, but it is quite a non-trivial proof that requires reasoning about integer arithmetic which is very hard to do in general.